source/beamline/BioXAS/BioXASSSRLMonochromatorRegionControl.cpp

```diff
@@ -0,0 +1,846 @@
+{
+	if (!crystalChangeCCWLimitStatus_ && crystalChangeCCWLimitStatus_ != limitStatus) {
+		crystalChangeCCWLimitStatus_ = limitStatus;
+		addChildControl(crystalChangeCWLimitStatus_);
+
+		connect( crystalChangeCCWLimitStatus_, SIGNAL(movingChanged(bool)), this, SIGNAL(movingChanged(bool)) );
+	}
+
+}
+
+void BioXASSSRLMonochromatorRegionControl::setRegionAStatusControl(AMControl *regionStatus)
+{
+	if (!regionAStatus_ && regionAStatus_ != regionStatus) {
+		regionAStatus_ = regionStatus;
+		addChildControl(regionAStatus_);
+
+		connect( regionAStatus_, SIGNAL(valueChanged(double)), this, SLOT(onRegionControlValueChanged()) );
+		connect( regionAStatus_, SIGNAL(movingChanged(bool)), this, SIGNAL(movingChanged(bool)) );
+	}
+}
+
+void BioXASSSRLMonochromatorRegionControl::setRegionBStatusControl(AMControl *regionStatus)
+{
+	if (!regionBStatus_ && regionBStatus_ != regionStatus) {
+		regionBStatus_ = regionStatus;
+		addChildControl(regionBStatus_);
+
+		connect( regionBStatus_, SIGNAL(valueChanged(double)), this, SLOT(onRegionControlValueChanged()) );
+		connect( regionBStatus_, SIGNAL(movingChanged(bool)), this, SIGNAL(movingChanged(bool)) );
+	}
+}
+
+void BioXASSSRLMonochromatorRegionControl::setValue(int newValue)
+{
+	if (value_ != newValue) {
+		value_ = newValue;
+		emit valueChanged(value_);
+	}
+}
+
+void BioXASSSRLMonochromatorRegionControl::setSetpoint(int newSetpoint)
+{
+	if (setpoint_ != newSetpoint) {
+		setpoint_ = newSetpoint;
+		emit setpointChanged(setpoint_);
+	}
+}
+
+void BioXASSSRLMonochromatorRegionControl::setMoveInProgress(bool isMoving)
+{
+	if (moveInProgress_ != isMoving) {
+		moveInProgress_ = isMoving;
+		emit movingChanged(moveInProgress_);
+
+		if (moveInProgress_)
+			emit moveStarted();
+	}
+}
+
+void BioXASSSRLMonochromatorRegionControl::onActionStepChanged(int stepIndex)
+{
+	emit moveStepChanged(stepDescription(stepIndex), stepInstruction(stepIndex), stepNotes(stepIndex));
+}
+
+void BioXASSSRLMonochromatorRegionControl::onRegionControlValueChanged()
+{
+	BioXASSSRLMonochromator::Region::State newRegion = BioXASSSRLMonochromator::Region::None;
+
+	if (regionAStatus_ && regionAStatus_->isConnected() && regionBStatus_ && regionBStatus_->isConnected()) {
+
+		int regionAVal = (int)regionAStatus_->value();
+		int regionBVal = (int)regionBStatus_->value();
+
+		if (regionAVal == BioXASSSRLMonochromator::Region::NotIn && regionBVal == BioXASSSRLMonochromator::Region::In)
+			newRegion = BioXASSSRLMonochromator::Region::B;
+		else if (regionAVal == BioXASSSRLMonochromator::Region::In && regionBVal == BioXASSSRLMonochromator::Region::NotIn)
+			newRegion = BioXASSSRLMonochromator::Region::A;
+	}
+
+	setValue(newRegion);
+}
+
+void BioXASSSRLMonochromatorRegionControl::onRegionChangeStarted()
+{
+	setMoveInProgress(true);
+}
+
+void BioXASSSRLMonochromatorRegionControl::onRegionChangeCancelled()
+{
+	moveCleanup(sender());
```

Comment:
Although incredibly handy, using ```sender()``` isn't a good object-oriented programming principle.  To quote the Qt docs:

> Warning: This function violates the object-oriented principle of modularity. However, getting access to the sender might be useful when many signals are connected to a single slot.
Warning: As mentioned above, the return value of this function is not valid when the slot is called via a Qt::DirectConnection from a thread different from this object's thread. Do not use this function in this type of scenario.

This behaviour can be completely reproduced using the QSignalMapper class.  It's more code, but it is more reliable and future-proof.